File: internal/services/container/nodekind/specs/purefunc.go
// internal/services/container/nodekind/specs/purefunc.go
package specs

import "yhbox/internal/services/container/nodekind"

// pureFunc builds a v4 §6 pure-function node spec. All have:
//   - No exec pins (IsPureData=true)
//   - Static data-in (passed in)
//   - Single data-out named "result"
//   - Defaults.literal echoes data-in zero values
func pureFunc(kind string, dataIn map[string]nodekind.PinType, resultType nodekind.PinType, defaults map[string]any) *nodekind.Spec {
	return &nodekind.Spec{
		Kind:       kind,
		Group:      "purefunc",
		DataIn:     dataIn,
		DataOut:    map[string]nodekind.PinType{"result": resultType},
		Defaults:   map[string]any{"literal": defaults},
		IsPureData: true,
	}
}

func init() {
	num2 := map[string]nodekind.PinType{"a": nodekind.PinNumber, "b": nodekind.PinNumber}
	any2 := map[string]nodekind.PinType{"a": nodekind.PinAny, "b": nodekind.PinAny}
	str2 := map[string]nodekind.PinType{"a": nodekind.PinString, "b": nodekind.PinString}
	zNum := map[string]any{"a": 0.0, "b": 0.0}
	zAny := map[string]any{}
	zStr := map[string]any{"a": "", "b": ""}

	// 算术 (6)
	nodekind.Register(pureFunc("Add", num2, nodekind.PinNumber, zNum))
	nodekind.Register(pureFunc("Sub", num2, nodekind.PinNumber, zNum))
	nodekind.Register(pureFunc("Mul", num2, nodekind.PinNumber, zNum))
	nodekind.Register(pureFunc("Div", num2, nodekind.PinNumber, zNum))
	nodekind.Register(pureFunc("Mod", num2, nodekind.PinNumber, zNum))
	nodekind.Register(pureFunc("Neg",
		map[string]nodekind.PinType{"x": nodekind.PinNumber},
		nodekind.PinNumber,
		map[string]any{"x": 0.0}))

	// 比较 (6) — Lt/LtEq/Gt/GtEq 是 number→bool, Eq/NotEq 是 any→bool
	nodekind.Register(pureFunc("Lt", num2, nodekind.PinBool, zNum))
	nodekind.Register(pureFunc("LtEq", num2, nodekind.PinBool, zNum))
	nodekind.Register(pureFunc("Gt", num2, nodekind.PinBool, zNum))
	nodekind.Register(pureFunc("GtEq", num2, nodekind.PinBool, zNum))
	nodekind.Register(pureFunc("Eq", any2, nodekind.PinBool, zAny))
	nodekind.Register(pureFunc("NotEq", any2, nodekind.PinBool, zAny))

	// 逻辑 (3) — pin type 是 bool 但 v3 expr 历史是 any, 先按 spec §6.3 用 bool
	bool2 := map[string]nodekind.PinType{"a": nodekind.PinBool, "b": nodekind.PinBool}
	zBool := map[string]any{"a": false, "b": false}
	zBoolT := map[string]any{"a": true, "b": true}
	nodekind.Register(pureFunc("And", bool2, nodekind.PinBool, zBoolT))
	nodekind.Register(pureFunc("Or", bool2, nodekind.PinBool, zBool))
	nodekind.Register(pureFunc("Not",
		map[string]nodekind.PinType{"x": nodekind.PinBool},
		nodekind.PinBool,
		map[string]any{"x": false}))

	// 字符串 (3)
	nodekind.Register(pureFunc("Concat", str2, nodekind.PinString, zStr))
	nodekind.Register(pureFunc("Contains",
		map[string]nodekind.PinType{"haystack": nodekind.PinString, "needle": nodekind.PinString},
		nodekind.PinBool,
		map[string]any{"haystack": "", "needle": ""}))
	nodekind.Register(pureFunc("Length",
		map[string]nodekind.PinType{"s": nodekind.PinString},
		nodekind.PinNumber,
		map[string]any{"s": ""}))

	// 转换 (3)
	nodekind.Register(pureFunc("ToString",
		map[string]nodekind.PinType{"x": nodekind.PinAny},
		nodekind.PinString,
		map[string]any{"x": ""}))
	nodekind.Register(pureFunc("ToNumber",
		map[string]nodekind.PinType{"x": nodekind.PinAny},
		nodekind.PinNumber,
		map[string]any{"x": 0.0}))
	nodekind.Register(pureFunc("ToBool",
		map[string]nodekind.PinType{"x": nodekind.PinAny},
		nodekind.PinBool,
		map[string]any{"x": false}))

	// Select 三元 (1)
	nodekind.Register(pureFunc("Select",
		map[string]nodekind.PinType{"cond": nodekind.PinBool, "a": nodekind.PinAny, "b": nodekind.PinAny},
		nodekind.PinAny,
		map[string]any{"cond": true}))
}

File: frontend/src/components/containers/nodeRegistry/specs/purefunc.ts
// frontend/src/components/containers/nodeRegistry/specs/purefunc.ts
// v4 §6 pure-function nodes (22): arithmetic / compare / logic / string / convert / select.
// Mirrors backend internal/services/container/nodekind/specs/purefunc.go.
//
// CRITICAL — type alignment with backend (v4 spec §6.2):
//   Lt/LtEq/Gt/GtEq/Eq/NotEq/And/Or/Not/Contains/ToBool → result = 'bool'
//   And/Or → data-in = 'bool'; Not → data-in = 'bool'
// (These differ from current pinSpec.ts which uses 'any' for those — backend is authoritative.)
import { register } from '../registry'
import type { NodeKindSpec, PinType } from '../index'

const purefuncVisualSuffix = { bg: 'bg-cyan-500/15', border: 'border-cyan-500/40' }

function pureFunc(args: {
  kind: string
  labelZh: string
  description?: string
  icon: string
  dataIn: Record<string, PinType>
  result: PinType
  literalDefaults: Record<string, unknown>
}): NodeKindSpec {
  return {
    kind: args.kind,
    group: 'purefunc',
    labelZh: args.labelZh,
    description: args.description ?? '',
    visual: { icon: args.icon, ...purefuncVisualSuffix },
    execIn: [],
    execOut: [],
    dataIn: args.dataIn,
    dataOut: { result: args.result },
    fields: [],
    defaults: { literal: args.literalDefaults },
    isPureData: true,
  }
}

// 算术 (6)
register(pureFunc({ kind: 'Add', labelZh: '加', icon: 'i-tabler-plus', dataIn: { a: 'number', b: 'number' }, result: 'number', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Sub', labelZh: '减', icon: 'i-tabler-minus', dataIn: { a: 'number', b: 'number' }, result: 'number', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Mul', labelZh: '乘', icon: 'i-tabler-x', dataIn: { a: 'number', b: 'number' }, result: 'number', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Div', labelZh: '除', icon: 'i-tabler-divide', dataIn: { a: 'number', b: 'number' }, result: 'number', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Mod', labelZh: '取余', icon: 'i-tabler-percentage', dataIn: { a: 'number', b: 'number' }, result: 'number', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Neg', labelZh: '取负', icon: 'i-tabler-arrow-down-right', dataIn: { x: 'number' }, result: 'number', literalDefaults: { x: 0 } }))

// 比较 (6) — backend says bool, not any
register(pureFunc({ kind: 'Lt', labelZh: '<', icon: 'i-tabler-math-lower', dataIn: { a: 'number', b: 'number' }, result: 'bool', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'LtEq', labelZh: '≤', icon: 'i-tabler-math-equal-lower', dataIn: { a: 'number', b: 'number' }, result: 'bool', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Gt', labelZh: '>', icon: 'i-tabler-math-greater', dataIn: { a: 'number', b: 'number' }, result: 'bool', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'GtEq', labelZh: '≥', icon: 'i-tabler-math-equal-greater', dataIn: { a: 'number', b: 'number' }, result: 'bool', literalDefaults: { a: 0, b: 0 } }))
register(pureFunc({ kind: 'Eq', labelZh: '==', icon: 'i-tabler-equal', dataIn: { a: 'any', b: 'any' }, result: 'bool', literalDefaults: {} }))
register(pureFunc({ kind: 'NotEq', labelZh: '!=', icon: 'i-tabler-equal-not', dataIn: { a: 'any', b: 'any' }, result: 'bool', literalDefaults: {} }))

// 逻辑 (3) — backend uses bool pins (v4 spec §6.3)
register(pureFunc({ kind: 'And', labelZh: '与', icon: 'i-tabler-ampersand', dataIn: { a: 'bool', b: 'bool' }, result: 'bool', literalDefaults: { a: true, b: true } }))
register(pureFunc({ kind: 'Or', labelZh: '或', icon: 'i-tabler-letter-o', dataIn: { a: 'bool', b: 'bool' }, result: 'bool', literalDefaults: { a: false, b: false } }))
register(pureFunc({ kind: 'Not', labelZh: '非', icon: 'i-tabler-exclamation-mark', dataIn: { x: 'bool' }, result: 'bool', literalDefaults: { x: false } }))

// 字符串 (3)
register(pureFunc({ kind: 'Concat', labelZh: '拼字符串', icon: 'i-tabler-plus-equal', dataIn: { a: 'string', b: 'string' }, result: 'string', literalDefaults: { a: '', b: '' } }))
register(pureFunc({ kind: 'Contains', labelZh: '包含', icon: 'i-tabler-search', dataIn: { haystack: 'string', needle: 'string' }, result: 'bool', literalDefaults: { haystack: '', needle: '' } }))
register(pureFunc({ kind: 'Length', labelZh: '长度', icon: 'i-tabler-ruler', dataIn: { s: 'string' }, result: 'number', literalDefaults: { s: '' } }))

// 转换 (3)
register(pureFunc({ kind: 'ToString', labelZh: '转字符串', icon: 'i-tabler-letter-t', dataIn: { x: 'any' }, result: 'string', literalDefaults: { x: '' } }))
register(pureFunc({ kind: 'ToNumber', labelZh: '转数字', icon: 'i-tabler-numbers', dataIn: { x: 'any' }, result: 'number', literalDefaults: { x: 0 } }))
register(pureFunc({ kind: 'ToBool', labelZh: '转布尔', icon: 'i-tabler-toggle-right', dataIn: { x: 'any' }, result: 'bool', literalDefaults: { x: false } }))

// Select 三元 (1)
register(pureFunc({ kind: 'Select', labelZh: '三元选择', icon: 'i-tabler-arrows-split', dataIn: { cond: 'bool', a: 'any', b: 'any' }, result: 'any', literalDefaults: { cond: true } }))
